PowerConnection, Animator components, components integration

dusho opened the pull request:
Introduces new components
- power connection (for all things that require or provide power)
- animator (for all things that animate)

Changes:
- added IPlugable interface , used in PowerNetwork instead of component, PowerConnection implements IPlugable

Generally introduced inter-component communication using furniture's params. 
Things to be still done:
- (DONE) fix the older PowerConnection UTs
- (DONE) way to declare/expose params in furniture.xml
- (DONE) furniture needs map with 'needs' for it to work - same as power icon, -> gas, temperature, input inventory, context selection, ... 
- (DONE) fix stylecop
- nice to have support for LUA functions instead of conditions in .xml (another separate PR)

Create WIP to gather some feedback if the implementation is going in right direction and whether there are better ideas how to handle or enhance things.
Most of things I tested seems to be working - I would ask other people to test it as well..

koosemose commented:
What do you mean about declaring/exposing params in xml, don't Components just use the furniture's Parameters?

dusho commented:
Yes, but currently they are defined in components source code. It just needs to be placed in .xml.. either in furniture element (that is already defined for other furnitures) or somehow define them in component element with some description about usage or something..

koosemose commented on Assets/Scripts/Models/Buildable/Components/PowerConnection.cs:
```diff
@@ -0,0 +1,208 @@
+            {
+                bool running = false;
+
+                // TODO: make adjustments in Parameters class to prevent double check for ContainsKey (indexer already checks it)
```
Please don't make that adjustment in Parameters, it is an intentional design that Parameters autocreates any parameter that is attempted to be accessed, and that requires it to check if it contains the key.

koosemose commented on Assets/Scripts/Models/Buildable/Components/PowerConnection.cs:
```diff
@@ -0,0 +1,208 @@
+                    running = FurnitureParams[CurIsRunningParamName].ToBool();
+                }                
+
+                return running;
```
This whole section can just simply be `return FurnitureParams[CurIsRunningParamName].ToBool()` it will autocreate it if it doesn't exist and the automatically created Parameter will already have a value of false (or a value that translates to false, anyways.

koosemose commented on Assets/Scripts/Models/Buildable/Components/PowerConnection.cs:
```diff
@@ -0,0 +1,208 @@
+            }
+
+            FurnitureParams.AddParameter(new Parameter(CurIsRunningParamName, false));
+
```
You don't need to explicitly AddParameter, due to the autocreation of Parameters when accessing a nonexistent one (in fact AddParameter just does `contents[parameter.GetName()] = parameter;`, it's an artifact from an older design and should have been removed), so `FurnitureParams[curAccumulatorChargeParamName] = 0f` would have exactly the same effect.

dusho commented:
- incorporated feedback from @koosemose 
- added ParameterDefinitions (not mandatory in .xml, but helpful for inter-component stuff)
- fixed UTs
- added IPlugable interface , used in PoweeNetwork instead of component, PowerConnection implements IPlugable 
- stylecop fixing

koosemose commented on Assets/StreamingAssets/Data/Furniture.xml:
```diff
@@ -284,10 +306,43 @@
             <Param name="burnTime" value="0" />
             <Param name="burnTimeRequired" value="60" />
```
Are these still used?

dusho replied:
no, it's not.. will remove those

koosemose commented:
You've got some stylecop issues, other than that, you look good to go.

dusho commented:
fixed stylecop

koosemose commented:
Oh good... now I can break my own PR all the sooner.. :)

koosemose commented:
Please replace usage of var with verbose typing, per style guidelines. And where possible try to use more verbose variable names (component vs. cmp).

koosemose commented on Assets/Scripts/Models/Buildable/Components/BuildableComponent.cs:
```diff
@@ -29,18 +34,49 @@ public BuildableComponent()
             get { return ParentFurniture.Parameters; }
         }
-        
+                        
```
Lots of extra spaces

koosemose commented on Assets/Scripts/Models/Buildable/Components/BuildableComponent.cs:
```diff
@@ -65,7 +101,7 @@ public static BuildableComponent Deserialize(XmlReader xmlReader)
             }
         }
-
+        
```
Trailing Spaces

koosemose commented on Assets/Scripts/Models/Buildable/Components/BuildableComponent.cs:
```diff
@@ -126,6 +154,38 @@ protected void InvokeContextMenuAction(Action<Furniture, string> function, strin
+            //// here evaluate all parameter conditions
+            if (conditions != null)
+            {
+                foreach (ParameterCondition cnd in conditions)
```
More verbose variable name

koosemose commented on Assets/Scripts/Models/Buildable/Components/PowerConnection.cs:
```diff
@@ -0,0 +1,259 @@
+            set
+            {
+                float oldAccPower = CurrentAccumulatorCharge.ToFloat();
+                if (oldAccPower != value)
```
I'm on the fence of rather or not this variable name needs changed, it may be enough that the purpose of the variable is clear from seeing just the name, but I'm not absolutely sure.

koosemose commented on Assets/Scripts/Models/Buildable/Components/Workshop.cs:
```diff
@@ -20,35 +20,92 @@ namespace ProjectPorcupine.Buildable.Components
+
+        public override bool CanFunction()
+        {
+            var curSetupChainName = CurrentProductionChainName.ToString();
```
Var Usage

koosemose commented on Assets/Scripts/Models/Buildable/Components/Workshop.cs:
```diff
@@ -161,18 +210,16 @@ protected override void Initialize()
         private void PlaceInventories(List<TileObjectTypeAmount> outPlacement)
         {
-            foreach (TileObjectTypeAmount outPlace in outPlacement)
+            foreach (var outPlace in outPlacement)
```
Var Usage

koosemose commented on Assets/Scripts/Models/Buildable/Components/Workshop.cs:
```diff
@@ -299,27 +337,28 @@ private void HaulingJobForInputs(ProductionChain prodChain)
                     if (desiredAmount > 0)
                     {
-                        Job job = new Job(
+                        Job jb = new Job(
```
More verbose variable name

koosemose commented on Assets/Scripts/Models/Buildable/Components/Workshop.cs:
```diff
@@ -299,27 +337,28 @@ private void HaulingJobForInputs(ProductionChain prodChain)
         private List<TileObjectTypeAmount> CheckForInventoryAtOutput(ProductionChain prodChain)
         {
-            List<TileObjectTypeAmount> outPlacement = new List<TileObjectTypeAmount>();
+            var outPlacement = new List<TileObjectTypeAmount>();
```
Var Usage

koosemose commented on Assets/Scripts/Models/Buildable/Components/Workshop.cs:
```diff
@@ -355,7 +394,7 @@ private List<TileObjectTypeAmount> CheckForInventoryAtOutput(ProductionChain pro
         private List<KeyValuePair<Tile, int>> CheckForInventoryAtInput(ProductionChain prodChain)
         {
-            List<KeyValuePair<Tile, int>> flaggedForTaking = new List<KeyValuePair<Tile, int>>();
+            var flaggedForTaking = new List<KeyValuePair<Tile, int>>();
```
Var Usage

koosemose commented on Assets/Scripts/Models/Buildable/Furniture.cs:
```diff
@@ -1090,42 +1067,44 @@ public IEnumerable<string> GetAdditionalInfo()
         {
-            string desc = component.GetDescription();
-            if (!string.IsNullOrEmpty(desc))
+            var desc = component.GetDescription();
```
Var Usage

koosemose commented:
Looks good, stylecop clean, works as advertised. Merged.
